CIPs/cip-120.md

```diff
@@ -0,0 +1,109 @@
+
+
+## Motivation
+<!--Motivation is critical for CIPs that want to change the Ceramic protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the CIP solves. CIP submissions without sufficient motivation may be rejected outright.-->
+Currently Ceramic uses one pubsub topic to propagate *update*, *query*, and *response* messages in the network. As the network grows the scalability of this approach will be limited. Separating messages into multiple topics doesn't neccessarily solve this problem entirely. However, it allows developers of less popular streams to not share the burden message propagation for more scaled applications.
+
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature.-->
+
+This CIP proposes a way to create a pubsub topic for any given stream given a *separator key* and a *separator value*. The *separator key* is defined by the `sep` field in the genesis event header and must be a string. The *separator value* is defined as the value of the field in the genesis header corresponding to the value of `sep`. The *separator value* is expected to be a byte string.
+
+All pubsub messages for streams that include a value for the `sep` field in the genesis header MUST use the following topic:
+
+```markdown
+/ceramic/<network-name>/<separator-key>/<separator-value>
+```
+
+- `network-name` - the name of the Ceramic network used
+- `separator-key` - the value of the `sep` separator key field
+- `separator-value` - the value of the field specified by the `sep` field, encoded as `base36`
+
+### Example
+
+If a stream has the following genesis header,
+
+```javascript
+{
+  header: {
+    sep: 'model',
+    model: <bytes-k2t6wyfsu4pg1hnttc39koxlw0pdn1kemzrn33cdfhrbooozvo7wb14sa1uqwi>
+  },
+  ...
+}
+```
+
+it would use the following pubsub topic on main net:
+
+```markdown
+/ceramic/mainnet/model/k2t6wyfsu4pg1hnttc39koxlw0pdn1kemzrn33cdfhrbooozvo7wb14sa1uqwi
+```
+
+### Streams without `sep`
+
+For backwards compatibility if a stream doesn't have a `sep` field or no value for the field specified by `sep`,  will use the following pubsub topic:
+
+```markdown
+/ceramic/<network-name>/
+```
+
+Which is the pubsub topic currently used by Ceramic.
+
+### Peer discovery
+
+While gossipsub used by Ceramic enables ambient discovery of peers, an initial set of peers already in a topic is required to recieve messages. It is therefore recommended that peers announce to the libp2p DHT that they are subscribed to a given pubsub topic. Peers can thus also discover other peers by looking up the DHT records for the given topic. The key that should be used for DHT providing and lookup MUST be `sha2-256(separator-key || separator-value)`.
```

Comment:
should the dht provide key have anything else to indicate it's for Ceramic specifically?  I guess the odds of someone random using the IPFS DHT to provide the hash of `"model || k2t6wyfsu4pg1hnttc39koxlw0pdn1kemzrn33cdfhrbooozvo7wb14sa1uqwi"` is pretty low, but I still wonder if we should do more to namespace ourselves in the global DHT

Reply:
The streamid is already namespaced to Ceramic since it includes the `0xce` multicodec.

Reply:
Made a change to add `network-name` so that different Ceramic networks don't overlap.